시나리오 3 : kafka를 활용해 DB I/O 진행

seongjunnoh opened the pull request:
## 요약
쿠폰 발급 요청이 대규모 트래픽으로 들어오는 상황을 가정할 때, 유저에게

1. 쿠폰 발급 성공 (200 OK)
2. 이미 발급받은 쿠폰이므로 쿠폰을 발급받을 수 없음 (400 error)
3. 쿠폰 재고가 없으므로 쿠폰을 발급받을 수 없음 (400 error)

3가지 중 하나를 빠르게 return 하기 위해서 쿠폰 발급 로직과 DB I/O와 관련된 코드를 분리

### 쿠폰 발급 로직을 담당하는 CouponService.issueCoupon 메서드 
- redis lua script를 활용해 현재 위의 1,2,3 중 어떤 상황인지 원자적으로 판단   
  - 2 or 3 에 해당하는 경우 CustomException을 throw 
  - 1에 해당하는 경우 redis에서의 쿠폰 재고 값을 DECR & DB에 UserCoupon save 와 Coupon의 재고정보 update를 위해 event 발행
  
### DB I/O 작업을 담당하는 CouponIssueConsumer.handleCouponIssue 메서드(= kafka listener)
- event를 수신하여 DB에 UserCoupon save & redis의 쿠폰 재고 값으로 Coupon 의 재고 정보를 update 시도
- KafkaConsumerConfig 에 kafka listener의 재시도 로직을 정의
  - userId, couponId 로 엔티티 조회 시 CustomException 이 발생할 경우, 재시도 X
  - UserCoupon save 중 user_coupon 에 설정된 (user, coupon) unique 제약조건에 의해 DataIntegrityViolationException 이 발생할 경우, warning log만 출력하고 현재 메시지 commit
    - 이미 동일한 (user, coupon) 쌍에 대한 UserCoupon 데이터가 DB에 저장된 상태이므로 추가적인 DB 쓰기 작업 없이 다음 메시지로 넘어가도 되므로
  - 그 외 모든 DB 쓰기 연산시에 발생하는 exception에 대해서는 최대 3회 재시도 (이후에는 DLT로 메시지 이동)
  - 전체 DB 쓰기 연산이 재시도되도록 handleCouponIssue 메서드에 @Transactional 설정

### 테스트 스크립트 수정
선착순 쿠폰 발급의 모니터링을 위해 쿠폰 발급 성공에 해당하는 로그 정보들을 조회한 후, 해당 로그의 '쿠폰 발급 요청 순번' 값이 오름차순 인지 검증하도록 코드 수정

### 부하테스트 결과
유저 50000명, 쿠폰 수량 3000개, 초당 10000회로 부하 테스트 진행 시, 정해진 쿠폰 수량만큼 쿠폰 발급이 성공하고 DB 에도 데이터들이 잘 들어가있는것을 확인했습니다
<img width="788" alt="image" src="https://github.com/user-attachments/assets/8fca64e8-a442-45d8-a049-b8cac79c752f" />
-> 빨간색 실패는 쿠폰 발급 성공 로그의 요청 순번 값들이 오름차순이 아니어서 발생한 것입니닷 (추후에 로그를 찍는 시점에 대해서도 고민해보아야 할 것 같습니다!)

<img width="1285" alt="image" src="https://github.com/user-attachments/assets/d5ef523c-14ff-4318-b590-7ade0491fd79" />

<img width="874" alt="image" src="https://github.com/user-attachments/assets/de8c6a61-91f8-477e-a204-d0a15f4fa03e" />

## 리뷰 요구사항
일단 쿠폰 발급을 담당하는 로직에서는 DB와 커넥션 없이 redis의 lua script 연산만을 통해 '유저의 쿠폰 발급 여부, 쿠폰 재고값' 을 알 수 있도록 하였고, DB 커넥션이 필요한 코드들은 모두 kafka event를 수신받아 비동기적으로 처리하도록 구현하였습니다!

하지만 이 방식은 DB가 아니라, redis에서의 유저-쿠폰 발급 여부, 쿠폰의 재고값을 바탕으로 유저에게 미리 응답하는 방식이라 저번에 진영님께서 언급해주셨던것처럼 실제 DB i/o 시에 문제가 발생하면 후처리를 어떻게 해야할지 애매해지는 것 같습니다.

DB 커넥션이 필요한 작업까지 모두 동기적으로 수행할 수 있으면서도, 쿠폰 발급 요청이 대규모 트래픽으로 들어오는 경우를 커버할 수 있는 방식에 대해서 좀 더 고민해보겠습니다! 혹시 생각하시는 다른 구현방식이 있으면 공유해주시면 감사하겠습니다!
(저번에 카프카 컨슈머를 여러개 띄워서 병렬적으로 처리하면 괜찮지 않나? 라고 말씀하신거 같긴한데 제가 제대로 못알아들은거 같습니다ㅎ)



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **신규 기능**
    - 쿠폰 발급 이벤트를 처리하는 Kafka 기반 비동기 아키텍처 도입
    - 쿠폰 재고 동기화를 위한 스케줄러 추가(5분마다 Redis와 DB 간 동기화)
    - 쿠폰 발급 이벤트 및 소비자, Kafka 설정 등 주요 이벤트 기반 클래스 추가

- **버그 수정**
    - 테스트 환경에서 Redis 포트 충돌 방지를 위한 동적 포트 할당 적용

- **테스트**
    - Kafka 및 Redis 환경을 반영한 통합 테스트 및 동시성 테스트 개선
    - 쿠폰 재고 동기화 검증 로직 명확화

- **환경설정/빌드**
    - 불필요한 Retry 관련 라이브러리 및 어노테이션 제거
    - 테스트용 Kafka, Redis 설정 및 consumer group 분리

- **기타**
    - 대규모 부하 테스트 파라미터 조정 및 검증 로직 개선
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

seongjunnoh committed:
chore : submodule update

seongjunnoh committed:
refactor : issueCoupon 메서드에서 DB I/O 작업 분리

- redis lua script에 user_coupon의 매핑관계도 저장하도록 수정
- 쿠폰 발급이 가능할 경우, DB I/O 작업을 위해 event publish

seongjunnoh committed:
feat : event consumer

- UserCoupon save 및 Coupon update를 하나의 트랜잭션으로 관리하는 handleCouponIssue 메서드 구현
- user_coupon의 unique 제약 조건을 제외한 DB 쓰기연산시 발생하는 exception에 대해 최대 3회 재시도 하도록 kafka config 추가

seongjunnoh committed:
refactor : test 코드

test 시에는 embeddedKafka 를 활용하도록 명시

seongjunnoh committed:
refactor : 테스트 스크립트 수정

- 선착순 쿠폰 발급 모니터링 코드 수정
- 쿠폰 발급 성공 로그의 '요청 순번'이 오름차순인지만 확인하도록 수정

pjy1368 commented on src/main/java/rediclaim/couponbackend/service/CouponIssueConsumer.java:
```diff
@@ -0,0 +1,75 @@
+    private final UserRepository userRepository;
+    private final CouponRepository couponRepository;
+
+    @KafkaListener(topics = "coupon-issue-events")
```
이거 아마.. coupons 일걸요? (production 기준 토픽 이름)

seongjunnoh replied:
넵 맞춰서 수정하겠습니다!

pjy1368 commented on src/main/java/rediclaim/couponbackend/service/CouponIssueConsumer.java:
```diff
@@ -0,0 +1,75 @@
+                            .build()
+            );
+
+            // Coupon 재고 update(redis의 쿠폰 재고 정보로 update)
+            String val = redisTemplate.opsForValue().get(stockKey);
+            if (val != null) {
+                int redisCount = Integer.parseInt(val);
+                coupon.setRemainingCount(redisCount);
+                couponRepository.save(coupon);
+            }
```
Redis 재고는 여기서 업데이트하지 않고 스케줄링 통해 일정 주기로 RDB에 동기화 맞춰주는 걸 생각했어요.
여기서 업데이트한다면 consume 실패했을 때 재고 다시 복원해야하니 복잡하지 않을까요?

pjy1368 commented on src/main/java/rediclaim/couponbackend/service/CouponIssueConsumer.java:
```diff
@@ -0,0 +1,75 @@
+            }
+
+            log.info("CouponIssueEvent 처리 완료 : userId={}, couponId={}", userId, couponId);
+        } catch (DataIntegrityViolationException e) {
+            log.warn("중복 CouponIssueEvent 수신 -> 무시합니다. : userId={}, couponId={}", userId, couponId);
+        } catch (Exception e) {
```
정확히 외래키 참조 무결성에 해당하는 Exception만 캐치해볼 수 있을까요? 코드 래빗 리뷰처럼 다른 DataIntegrityViolationException 에러가 잡힐 수 있어 보여서요~

pjy1368 commented on src/main/java/rediclaim/couponbackend/service/CouponService.java:
```diff
@@ -26,86 +28,76 @@
-        if (remaining < 0) {        // 해당 쿠폰 재고 없음
+        String issuedUsersKey = ISSUED_USERS_KEY_PREFIX + couponId;
+
+        Long remaining = allocateStockAtomically(stockKey, issuedUsersKey, userId);
+        if (remaining == -2) {
```
오 멱등키 쓰셨군요👍

pjy1368 commented on src/main/java/rediclaim/couponbackend/service/CouponService.java:
```diff
@@ -26,86 +28,76 @@
+
+        Message<CouponIssueEvent> message = MessageBuilder
+                .withPayload(event)
+                .setHeader(KafkaHeaders.TOPIC, "coupon-issue-events")       // 토픽 설정
```
```suggestion
                .setHeader(KafkaHeaders.TOPIC, "coupons")       // 토픽 설정
```

seongjunnoh replied:
토픽이름 이렇게 수정하겠습니다!

pjy1368 commented on src/test/java/rediclaim/couponbackend/service/CouponServiceInMultiThreadTest.java:
```diff
@@ -89,6 +95,9 @@ void should_issue_exact_quantity_in_multi_thread() throws Exception {
         latch.await();
         executor.shutdown();
 
+        // consumer가 DB i/o 작업을 마칠때까지 잠깐 wait
+        Thread.sleep(1000);
```
이걸 왜 해야하나요? DB i/o는 블로킹이어서 카프카 컨슘 스레드가 어차피 작업 마칠 떄까지 블로킹됩니다.

seongjunnoh replied:
현재 테스트 코드 상 테스트의 메인 스레드와 카프카 컨슘 스레드가 분리되어 있고, 메인 스레드는 카프카 컨슘 스레드가 DB i/o 작업을 마쳤는지를 확인하지 않고 바로 DB에 저장된 쿠폰의 재고정보를 확인하고 있어서 Thread.sleep(1000) 을 통해 메인 스레드를 잠시 멈춰서 소비자 스레드가 DB i/o를 끝낼 시간을 벌어주는 꼼수를 써봤습니다,,ㅎ
코드래빗 리뷰 참고해서 코드 수정해보겠습니다!

pjy1368 reviewed (commented):
[다음 스텝 고민]
- 사용자가 동기적으로 쿠폰 발급 사실을 알 수 있도록 하기
- CouponService에서 KafkaProduce를 하고 있는데, 이 메서드 내부에서 kafkaProduct한 이벤트를 곧바로 kafkaConsume할 수 있는지 리서치하기
- 즉, 메서드 내부에서 `kafkaProducer.publish(event); kafkaConsumer.consume(event);` 이런 형태가 가능한지 조사하면 됨.
- 이렇게 하면 기존의 카프카 안 쓴 버전이랑 똑같은 것 아닌가?
    - 카프카 컨슈머 스레드는 n개로 정해져 있으므로 메서드 내부에서 kafkaConsumer.consume 한다고 해서 곧바로 consume 되지 않을 수 있음. (이미 컨슈머 스레드가 동작하고 있다면)
    - 즉, 처리량이 일정 부분 버퍼링되므로 DB 부하를 줄일 수 있음.
    - 대신 사용자가 너무 오래 기다리면 request timeout 발생할 여지는 있음. (이건 테스트로 검증)

seongjunnoh committed:
refactor : test 코드 수정

- awaitility 의존성 추가
- 단순히 thread sleep가 아니라, kafka consume 스레드가 DB i/o 작업을 마칠때까지 기다리도록 코드 수정

seongjunnoh committed:
refactor : kafka topic 네이밍 수정

seongjunnoh committed:
refactor : issueCoupon 메서드에 kafka 메시지 전송 실패 상황에 대한 예외처리 추가

kafka 브로커가 메시지 수신에 실패하는 경우에 대하여 redis의 정보를 복원

seongjunnoh committed:
refactor : kafka consumer는 UserCoupon 만을 save 하도록 수정

UserCoupon의 unique 제약 조건은 무시 & 다른 exception에 대해서는 throw -> 재시도 후 DLT로 가도록 수정

seongjunnoh committed:
feat : redis의 쿠폰 재고 정보 -> RDB의 쿠폰 재고 정보의 동기화를 위한 스케줄러

seongjunnoh committed:
refactor : 테스트 스크립트

teardown 메서드 timeout 옵션 추가, RPS값 수정

seongjunnoh committed:
refactor : 테스트 코드 수정

redis 쿠폰 재고값 -> DB 쿠폰 재고값의 update를 테스트 시에 수동으로 실시하도록 코드 수정

seongjunnoh committed:
refactor : 사용하지 않는 의존성 삭제

seongjunnoh reviewed (commented):
진영님과 코드 래빗 리뷰 참고하여 코드 수정 하였습니다!

수정 사항은 아래와 같습니다.

1. CouponService.issueCoupon 메서드에서 kafka 메시지를 send 하는 과정에서, send 실패시에 대한 예외처리 코드 추가
2. CouponIssueConsumer.handleCouponIssue 메서드는 kafka 메시지를 수신받아 UserCoupon 만을 DB에 save 하도록 코드 수정
3. redis의 쿠폰 재고값과 DB의 쿠폰 재고값의 동기화는 스케줄러를 통하여 이루어지도록 코드 수정

그리고 유저 수 10만, 쿠폰 수량 1만으로 설정한 후 초당 15000 만큼 요청보내도록 테스트 스크립트를 수정한 후, 로컬에서 테스트를 여러번 진행하였는데, 쿠폰 수량과 유저 수가 크고 RPS가 높아서 그런지 timeout 발생이슈로 200 OK respons의 count 값이 1만보다 작게 측정되는 경우가 종종 발생하긴 합니다!
<img width="828" alt="image" src="https://github.com/user-attachments/assets/172cc24e-ac2e-4424-acec-6c79a9659f11" />

(DB와 redis의 값은 모두 문제없이 잘 들어간 것 확인했습니다!)

seongjunnoh committed:
refactor : syncCouponStock 메서드의 try-catch 구문

모든 Exception을 catch 할 수 있도록 코드 수정
